### DISCUSSION TITLE
Enabling Concurrent Mode in NextJS

### DISCUSSION THREAD

**Discussion by cyrus-za** [vercel/next.js#10645] [Help] [answered]
Via: https://github.com/zeit/next.js/issues/10643

I'd like to test out concurrent mode in my current nextjs project, now that its been released in experimental mode. The way to enable it in a normal react app is quite straght forward by switching `ReactDom.render(<App />, rootEl)` to `ReactDom.createRoot(rootEl).render(<App />)`

The problem is nextjs does that part for each page under the hood. It would be nice to have a option in next.config.js to turn concurrent mode on, or even better a way to opt-in per page (e.g. if there are a few old legacy pages that will not work in react strict mode, those dont get opted in, but we can opt in the ones that are ready for this).

**Comment by timneutkens** [comment, marked as answer]
You can enable it as experimental under:

```
// next.config.js

module.exports = {
  experimental: {
    reactMode: 'concurrent'
  }
}
```

There is no way to do per-page opt-in as the React tree can't be rendered in two different modes (also the entrypoint is different).

So to start you should enable strict mode and make all pages are compatible:

```
// next.config.js
module.exports = {
  reactStrictMode: true
}
```

**Reply by salazarm** [reply]
Is there a way to opt out in Next.js 13? We're upgrading to Next 13 with a static export for now, I'd like to tackle the concurrent mode separately (We tried enable concurrent mode before but there were bugs in libraries we were using). It would be nice to have an incremental path for upgrading :)

**Reply by controversial** [reply]
@salazarm No; every major feature of the Next.js 13 App Router is built using concurrency / suspense

**Reply by salazarm** [reply]
@controversial I understand, but I'm using the pages router for now as we incrementally adopt the latest features.

**Reply by salazarm** [reply]
I guess I should stick to nextjs 12?

**Reply by salazarm** [reply]
Nextjs13 has other good features that don't involve the AppRouter, like `transpilePackages`, performance and memory improvements, etc. So it'd be nice to be able to use next.js 13 with the pages router and without React concurrent mode as an incremental upgrade path. Please consider it :)

**Reply by controversial** [reply]
I’m not sure whether there’s a way to use the Next.js 13 pages router without concurrent mode; you should open a new discussion for visibility.

**Reply by rizwanaslambhatti737-cmyk** [reply]
next.config.js
<img width="720" height="1600" alt="Screenshot_20260718-003046" src="https://github.com/user-attachments/assets/d58fdcf0-c255-4f08-82fe-0bd96fdb8d73" />

**Reply by rizwanaslambhatti737-cmyk** [reply]
Nice to see place for business started

**Comment by controversial** [comment]
I can't seem to enable concurrent mode without breaking all client-side React functionality. In my own app, when I set `experimental.reactMode` to either "blocking" or "concurrent," my app still renders server-side, but React never mounts on the client side, so all interactivity is gone. I reproduced this issue with more minimal apps from the included examples, as well—enabling concurrent mode seems to break React on the client side altogether. I'm not sure if I'm doing something wrong, or whether this is just the stage that the experimental concurrent mode support is currently at. What could cause this?

**Reply by devknoll** [reply]
Are you replacing your `react-dom` and `react` versions in `package.json` with `0.0.0-experimental-aae83a4b9`?

**Reply by ghost** [reply]
I am facing the same issues. I changed those versions too.

**Reply by devknoll** [reply]
@controversial Could you try the suggestion in https://github.com/zeit/next.js/issues/11405 and see if it addresses your issue too?

**Reply by controversial** [reply]
I had replaced `react` with the `experimental` version but not `react-dom`. Replacing both fixes my problem. Thanks so much!

**Comment by EricKit** [comment]
I can't get this to work anymore with `React 17.0.0-rc.2`, it produces the error 

> Uncaught TypeError: _reactDom.default.unstable_createRoot is not a function

Which doesn't make sense to me because in React 17, `unstable_createRoot` should be right.

**Reply by hosmelq** [reply]
You have to use the experimental.

yarn add react@experimental react-dom@experimental

**Reply by EricKit** [reply]
Makes complete sense I moved off of experimental, I assumed it'd be included in the RC for 17... bad assumption, thanks!

**Comment by enes-get-local** [comment]
I run my next project at concurrent mode but it is super slow now. That caused since it is still experiemental version or i need to configure something else ?

Basically added these options to next.config.js :

```
experimental: {
    reactRoot: 'concurrent',
    reactStrictMode: true
  }
```

dependencies here :

```
"react": "^0.0.0-experimental-63927e084",
"react-dom": "^0.0.0-experimental-63927e084",
```

Now project is super jittery and slow

**Reply by saminnet** [reply]
`reactStrictMode: true` needs to be in outer scope of `experimental`, like this:
```
reactStrictMode: true,
experimental: {
  reactRoot: 'concurrent',
},
```
